### PR TITLE
Use semaphore for queue lock

### PR DESCRIPTION
**Pull request by jkemp101**
- [x] Add tests for max workers > 1
- [ ] Test Redis impact with large number of locked queues, large number of workers, and new tasks being queued to locked queues

### CODE REVIEW

File: tasktiger/redis_semaphore.py [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+        with open(os.path.join(os.path.dirname(os.path.realpath(__file__)),
+                               'lua/semaphore.lua')) as f:
+            self._semaphore = self.redis.register_script(f.read())
+
+    def release(self):
+        """Release semaphore."""
+
+        self.redis.zrem(self.name, self.id)
+
+    def acquire(self):
```
Comment (tsx) [inline tasktiger/redis_semaphore.py]: We could add an optional `def acquire(self, time=time.time):` parameter, pass time explicitly in tests and avoid freezefrog dependency.

File: tasktiger/worker.py [hunk 2]
```diff
@@ -353,24 +367,26 @@ def _get_queue_lock(self, queue, log):
...
+                max_workers = 1
                 break
 
         # Single worker queues require us to get a queue lock before
         # moving tasks
-        if single_worker_queue:
-            queue_lock = Lock(self.connection, self._key('qlock', queue),
-                              timeout=self.config['ACTIVE_TASK_UPDATE_TIMEOUT'])
-            acquired = queue_lock.acquire(blocking=False)
+        if max_workers:
```
Comment (thomasst) [inline tasktiger/worker.py]: Update "Single worker queues" comment above.

File: tasktiger/__init__.py [hunk 3]
```diff
@@ -150,6 +151,11 @@ def __init__(self, connection=None, config=None, setup_structlog=False):
             # locking techniques.
             'SINGLE_WORKER_QUEUES': [],
 
+            # The maximum number of workers that will be allowed to actively
+            # process tasks for any queue. Similar to Single Worker Queues
+            # but allows specifying any number 1 or greater.
+            'MAX_WORKERS_PER_QUEUE': None,
```
Comment (thomasst) [inline tasktiger/__init__.py]: Should there be  an option to configure this per queue, and then deprecate `SINGLE_WORKER_QUEUES`? E.g.:
```
MAX_WORKERS_PER_QUEUE = {
    None: 4,    # default
    'some-queue': 1,
    'some-queue.subqueue': 2,
    'other-queue': 6,
}
```

File: tasktiger/__init__.py [hunk 4]
```diff
@@ -323,6 +331,25 @@ def delay(self, func, args=None, kwargs=None, queue=None,
 
         return task
 
+    def queue_system_lock(self, queue, timeout):
```
Comment (thomasst) [inline tasktiger/__init__.py]: Is this used anywhere?

File: tasktiger/worker.py [hunk 5]
```diff
@@ -11,16 +11,20 @@
...
 
 from ._internal import *
 from .exceptions import RetryException, TaskNotFound
+from .redis_semaphore import Semaphore
 from .retry import *
 from .stats import StatsThread
 from .task import Task
 from .timeouts import UnixSignalDeathPenalty, JobTimeoutException
 
+LOCK_REDIS_KEY = 'qlock'
```
Comment (thomasst) [inline tasktiger/worker.py]: Should consider adding this to the "Redis keys" comment section in `tasktiger/__init__.py`.

File: tasktiger/__init__.py (hunk 3 above)
Comment (jkemp101) [reply tasktiger/__init__.py]: I'm a bit torn on how we should handle queue configuration as we add more and more queue specific options. Right now a queue's configuration is spread across multiple keys in the config. Do you think we should consider moving to something like this instead and stop using `ONLY_QUEUES`, `BATCH_QUEUES`, `SINGLE_WORKER_QUEUES`/`MAX_WORKERS_PER_QUEUE`?
```
QUEUES = {
    'a': {
        max_workers: 5,
        batch: 50
    },
    'b': {}
}
```
We could also make it smart enough to detect if `ONLY_QUEUES` is a list or a dictionary so we don't have to add `QUEUES` to the config. 

If we wanted to make that change I could remove `MAX_WORKERS_PER_QUEUE` for now and just support it as a command line option until we refactor the config handling.

**Pull request merged**